File: apps/ios/Core/MandarinPinyin.swift
import Foundation

// Builds on richardguerre's Mandarin contribution in Chuloo/mural#4.
public struct MandarinPronunciationToken: Equatable, Sendable {
    public let text: String
    public let pinyin: String?
    public init(text: String, pinyin: String?) {
        self.text = text
        self.pinyin = pinyin
    }
}

public enum MandarinPinyin {
    /// Uses the system's word readings so 银行 and 旅行 keep different readings of 行.
    /// Punctuation, spacing and unrecognized characters remain exactly as supplied.
    public static func tokens(_ text: String) -> [MandarinPronunciationToken] {
        guard !text.isEmpty else { return [] }
        let source = text as NSString
        let tokenizer = CFStringTokenizerCreate(nil, text as CFString,
            CFRange(location: 0, length: source.length), kCFStringTokenizerUnitWord,
            CFLocaleCreate(nil, CFLocaleIdentifier("zh_CN" as CFString)))!
        var result: [MandarinPronunciationToken] = []
        var cursor = 0
        while CFStringTokenizerAdvanceToNextToken(tokenizer).rawValue != 0 {
            let range = CFStringTokenizerGetCurrentTokenRange(tokenizer)
            guard range.location >= cursor, range.length > 0, range.location + range.length <= source.length else { continue }
            if range.location > cursor {
                result.append(.init(text: source.substring(with: NSRange(location: cursor, length: range.location - cursor)), pinyin: nil))
            }
            let word = source.substring(with: NSRange(location: range.location, length: range.length))
            var reading: String?
            if containsHan(word), let latin = CFStringTokenizerCopyCurrentTokenAttribute(tokenizer, kCFStringTokenizerAttributeLatinTranscription) as? String {
                // Apple's dictionary uses v for ü in some readings (e.g. 旅行).
                let normalized = String(String.UnicodeScalarView(latin.lowercased().unicodeScalars.map { $0 == "v" ? "ü" : $0 }))
                    .precomposedStringWithCanonicalMapping.trimmingCharacters(in: .whitespacesAndNewlines)
                if !normalized.isEmpty && !containsHan(normalized) { reading = normalized }
            }
            result.append(.init(text: word, pinyin: reading))
            cursor = range.location + range.length
        }
        if cursor < source.length {
            result.append(.init(text: source.substring(from: cursor), pinyin: nil))
        }
        return result
    }

    /// A separate reading aid; source text and learning evidence are never replaced.
    public static func reading(_ text: String) -> String? {
        let parts = tokens(text)
        guard parts.contains(where: { $0.pinyin != nil }) else { return nil }
        var result = ""
        for part in parts {
            let value = part.pinyin ?? part.text
            if let last = result.last, let first = value.first,
               (last.isLetter || last.isNumber), (first.isLetter || first.isNumber) {
                result.append(" ")
            }
            result.append(value)
        }
        return result
    }

    private static func containsHan(_ text: String) -> Bool {
        text.unicodeScalars.contains { scalar in
            switch scalar.value {
            case 0x3400...0x4DBF, 0x4E00...0x9FFF, 0xF900...0xFAFF,
                 0x20000...0x2FA1F, 0x30000...0x3347F: true
            default: false
            }
        }
    }
}

public struct CaptionSegment: Equatable, Sendable {
    public let text: String
    public let lookup: String?
}

public enum CaptionWords {
    /// Keeps every source character, linking Chinese words instead of whole sentences.
    public static func segments(_ text: String, languageID: String) -> [CaptionSegment] {
        if languageID == "zh" {
            return MandarinPinyin.tokens(text).map {
                CaptionSegment(text: $0.text, lookup: $0.text.contains(where: \.isLetter) ? $0.text : nil)
            }
        }
        var result: [CaptionSegment] = []
        var run = ""
        for character in text {
            if let last = run.last, last.isWhitespace != character.isWhitespace {
                result.append(segment(run)); run = ""
            }
            run.append(character)
        }
        if !run.isEmpty { result.append(segment(run)) }
        return result
    }

    private static func segment(_ text: String) -> CaptionSegment {
        let word = text.trimmingCharacters(in: .punctuationCharacters.union(.whitespacesAndNewlines))
        return CaptionSegment(text: text, lookup: word.contains(where: \.isLetter) ? word : nil)
    }
}

File: apps/android/app/src/main/java/chat/mural/core/MandarinPinyin.kt
package chat.mural.core

import java.text.Normalizer

data class MandarinPronunciationToken(val text: String, val pinyin: String?)

/** Word segmentation and readings for Han text, supplied by the platform. */
interface HanReader {
    /** Splits text into consecutive pieces whose concatenation is the original text. */
    fun words(text: String): List<String>

    /** A Latin transcription of one word, or null when the platform has none. */
    fun reading(word: String): String?

    /** Pronunciation may use a phrase spanning several caption words without changing their tap targets. */
    fun pronunciationTokens(text: String): List<MandarinPronunciationToken> =
        words(text).map { MandarinPronunciationToken(it, if (MandarinPinyin.containsHan(it)) reading(it) else null) }
}

object MandarinPinyin {
    @Volatile var reader: HanReader? = null

    /** Punctuation, spacing and unrecognized characters remain exactly as supplied. */
    fun tokens(text: String, reader: HanReader? = this.reader): List<MandarinPronunciationToken> {
        if (text.isEmpty()) return emptyList()
        val tokens = reader?.pronunciationTokens(text)?.takeIf { it.joinToString("") { part -> part.text } == text }
            ?: return listOf(MandarinPronunciationToken(text, null))
        return tokens.map { part ->
            val latin = part.pinyin?.let(::normalize)
            part.copy(pinyin = latin?.takeIf { it.isNotEmpty() && !containsHan(it) })
        }
    }

    /** Reader segmentation whose pieces reconstruct the text, or null when the platform offers none. */
    fun words(text: String, reader: HanReader? = this.reader): List<String>? =
        reader?.words(text)?.takeIf { it.joinToString("") == text }

    /** A separate reading aid; source text and learning evidence are never replaced. */
    fun reading(text: String, reader: HanReader? = this.reader): String? {
        val parts = tokens(text, reader)
        if (parts.none { it.pinyin != null }) return null
        val result = StringBuilder()
        for (part in parts) {
            val value = part.pinyin ?: part.text
            if (result.isNotEmpty() && value.isNotEmpty() &&
                Character.isLetterOrDigit(result.codePointBefore(result.length)) && Character.isLetterOrDigit(value.codePointAt(0))) result.append(' ')
            result.append(value)
        }
        return result.toString()
    }

    fun containsHan(text: String): Boolean = text.codePoints().anyMatch { cp ->
        cp in 0x3400..0x4DBF || cp in 0x4E00..0x9FFF || cp in 0xF900..0xFAFF || cp in 0x20000..0x2FA1F || cp in 0x30000..0x3347F
    }

    /** Dictionary readings may spell ü as v (旅行); readings are lowercase and precomposed. */
    private fun normalize(latin: String): String =
        Normalizer.normalize(latin.lowercase().replace('v', 'ü'), Normalizer.Form.NFC).trim()
}

object HanWords {
    /** Joins adjacent segments that together form a known word, longest match first, up to four segments. */
    fun merge(segments: List<String>, words: Set<String>): List<String> {
        val result = mutableListOf<String>()
        var i = 0
        while (i < segments.size) {
            var taken = 1
            var chosen = segments[i]
            for (count in minOf(4, segments.size - i) downTo 2) {
                val joined = segments.subList(i, i + count).joinToString("")
                if (joined in words) { taken = count; chosen = joined; break }
            }
            result += chosen; i += taken
        }
        return result
    }
}

data class CaptionSegment(val text: String, val lookup: String?)

object CaptionWords {
    /** Keeps every source character, linking Chinese words instead of whole sentences. */
    fun segments(text: String, languageID: String, reader: HanReader? = MandarinPinyin.reader): List<CaptionSegment> {
        if (languageID == "zh") {
            if (text.isEmpty()) return emptyList()
            val words = MandarinPinyin.words(text, reader) ?: return listOf(CaptionSegment(text, null))
            return words.map { CaptionSegment(it, it.takeIf(::containsLetter)) }
        }
        val result = mutableListOf<CaptionSegment>()
        val run = StringBuilder()
        for (character in text) {
            if (run.isNotEmpty() && run.last().isWhitespace() != character.isWhitespace()) {
                result += segment(run.toString()); run.clear()
            }
            run.append(character)
        }
        if (run.isNotEmpty()) result += segment(run.toString())
        return result
    }

    private fun segment(text: String): CaptionSegment {
        val word = text.trim { it.isWhitespace() || isPunctuation(it) }
        return CaptionSegment(text, word.takeIf(::containsLetter))
    }

    private fun containsLetter(text: String): Boolean = text.codePoints().anyMatch { Character.isLetter(it) }

    private fun isPunctuation(c: Char): Boolean = when (Character.getType(c).toByte()) {
        Character.CONNECTOR_PUNCTUATION, Character.DASH_PUNCTUATION, Character.START_PUNCTUATION, Character.END_PUNCTUATION,
        Character.INITIAL_QUOTE_PUNCTUATION, Character.FINAL_QUOTE_PUNCTUATION, Character.OTHER_PUNCTUATION -> true
        else -> false
    }
}
